src/serialization/serialization.cpp

```diff
@@ -0,0 +1,153 @@
+    uint32_t crc = 0xFFFFFFFFU;
+    const uint8_t* bytes = static_cast<const uint8_t*>(data);
+    for (size_t i = 0; i < length; ++i) {
+        crc ^= bytes[i];
+        for (int j = 0; j < 8; ++j) {
+            if (crc & 1)
+                crc = (crc >> 1) ^ 0xEDB88320U;
+            else
+                crc = crc >> 1;
+        }
+    }
+    return crc ^ 0xFFFFFFFFU;
+}
+
+/// Serializes a DataFile object to an output stream.
+/// The number of positions in header must match the number of positions in entry.
+void serialize(const DataFile& data, std::ostream& stream) {
+    DataFileHeader header = data.header;
+    hton(header);
+    header.crc = calculateCRC32(&header, sizeof(header) - sizeof(header.crc));
+    header.crc = htonl(header.crc);
+    stream.write(reinterpret_cast<const char*>(&header), sizeof(header));
+
+    EulerAngles relative_attitude = data.relative_attitude;
+    hton(relative_attitude);
+    stream.write(reinterpret_cast<const char*>(&relative_attitude), sizeof(relative_attitude));
+
+    for (uint32_t i = 0; i < data.header.num_positions; ++i) {
+        LocationRecord record = data.positions[i];
+        hton(record);
+        stream.write(reinterpret_cast<const char*>(&record), sizeof(LocationRecord));
+    }
+}
+
+/// Deserializes a DataFile object from an input stream.
+DataFile deserialize(std::istream& stream) {
+    DataFile data;
+    data.header = readHeader(stream);
+
+    stream.read(reinterpret_cast<char*>(&data.relative_attitude), sizeof(EulerAngles));
+    if (stream.gcount() != sizeof(EulerAngles)) {
+        throw std::ios_base::failure("Failed to read relative_attitude");
+    }
+    ntoh(data.relative_attitude);
+
+    data.positions = std::make_unique<LocationRecord[]>(data.header.num_positions);
+    for (uint32_t i = 0; i < data.header.num_positions; ++i) {
+        stream.read(reinterpret_cast<char*>(&data.positions[i]), sizeof(LocationRecord));
+        if (stream.gcount() != sizeof(LocationRecord)) {
+            throw std::ios_base::failure("Failed to read location record");
+        }
+        ntoh(data.positions[i]);
+    }
+
+    return data;
+}
+
+
+/// Validates the magic number in the header.
+bool isValidMagicNumber(const char magic[4]) {
+    return magic[0] == 'F' && magic[1] == 'O' && magic[2] == 'U' && magic[3] == 'N';
+}
+
+DataFileHeader readHeader(std::istream& stream) {
+    DataFileHeader header;
+    stream.read(reinterpret_cast<char*>(&header), sizeof(DataFileHeader));
+    if (stream.gcount() != sizeof(header)) {
+        throw std::ios_base::failure("Failed to read header");
+    }
+    if (!isValidMagicNumber(header.magic)) {
+        throw std::ios_base::failure("Invalid magic number in header");
+    }
+
+    // Validate CRC
+    uint32_t expected_crc = calculateCRC32(&header, sizeof(header) - sizeof(header.crc));
+    if (ntohl(header.crc) != expected_crc) {
+        std::cerr << "Expected CRC: " << expected_crc << ", Found CRC: " << ntohl(header.crc) << std::endl;
```

Comment:
Replace this with `LOG_ERROR("Expected CRC: " << expected_crc << ", Found CRC: " << ntohl(header.crc));`. You will need to `#include "common/logging.hpp"`